README.rst
========
PapeRace
========

Computer implementation of the game, which could be played on a piece of
checkered paper.


Description
===========

This is my first game written in Python. Keep that in mind and be forgiving,
please ;) All comments in the source code are in Polish - sorry for that!

The goal of the game is to beat one lap on the racetrack by placing dots at
the intersection of the grids. The player should move clockwise. The first dot
can be placed at a distance of one grille (directly on the right or diagonally).
Consider this movement as a vector [x, y], where 'x' is a component of
the right/left movement, and 'y' is a component of the up and down motion. Each
next move should be the same as the previous one (same motion vector), but each
component can be freely modified by + 1/-1. In other words, you can move to
the field by distance [x, y] from the current position OR to one of the 8
adjacent fields. This simulates the acceleration/deceleration of the vehicle
and torsion. The player loses when he is out of the track or out of the game
window.

  .. image:: https://github.com/mariomoore/paperace/blob/master/src/paperace/plansza/plansza.png

Features
========

- The game is inteded for 1-6 human players
- At the beginning of each race the player's position on the red starting line is random
- In each race the player's color is random
- The color of the current player is marked with a border with the same color around the racetrack


Note
====

This project has been set up using PyScaffold 3.0.3. For details and usage
information on PyScaffold see http://pyscaffold.org/.


CHANGELOG.rst
=========
Changelog
=========

Version 1.0
===========

- First release


tests/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for paperace.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""
from __future__ import print_function, absolute_import, division

# import pytest


tests/test_car.py
import pytest
from paperace.car import Car
from paperace.const import BLACK

def test_car_init():
    """Sprawdzenie, czy objekt Car jest prawidłowo inicjalizowany"""
    car = Car(60, 60, BLACK)
    assert car.xpos == 60
    assert car.ypos == 60
    assert car.xv == 0
    assert car.yv == 0
    assert car.color == BLACK

def test_car_move():
    """Sprawdzanie czy pojazd porusza się prawidłowo w dwóch kierunkach"""
    car = Car(60, 60, BLACK)
    car.move((80, 80))
    assert car.xpos == 80
    assert car.ypos == 80
    assert car.xv == 20
    assert car.yv == 20
    car.move((60, 60))
    assert car.xpos == 60
    assert car.ypos == 60
    assert car.xv == -20
    assert car.yv == -20


tests/test_paperace.py
import pytest
from paperace.__main__ import get_nearest_point
import pygame
from paperace.__main__ import check_crossing_lines
from paperace.__main__ import init_cars_rand
import os
import sys

@pytest.mark.parametrize('mousexy, answerxy', [
    ((0, 0), (0, 0)),
    ((800, 600), (800, 600)),
    ((19, 84), (20, 80)),
    ((50, 95), (60, 100)),
])
def test_get_nearest_point(mousexy, answerxy):
    """Sprawdza czy metoda podaje najbliższy punkt (przecięcie kratki) po
    kliknięciu myszą"""
    point = get_nearest_point(mousexy)
    assert point == answerxy

@pytest.mark.parametrize('begin, end, colorscr, colorbgd, answer', [
    ((0, 0), (10, 10), (0, 0, 0, 255), (0, 0, 255, 255), True), # BLACK, BLUE
    ((10, 0), (20, 10), (0, 0, 0, 255), (0, 0, 255, 255), False), # BLACK, BLUE
    ((20, 0), (30, 10), (255, 0, 0, 255), (0, 0, 255, 255), True), # RED, BLUE
    ((0, 0), (10, 0), (0, 0, 0, 255), (0, 0, 255, 255), True), # BLACK, BLUE
    ((0, 0), (0, 10), (0, 0, 0, 255), (0, 0, 255, 255), True), # BLACK, BLUE
])
def test_check_crossing_lines(begin, end, colorscr, colorbgd, answer):
    """Sprawdza czy metoda prawidłowo wskazuje przecięte linie o zadanych
    kolorach"""
    screen = pygame.image.load(os.path.join(sys.path[0], 'poziome.png'))
    background = pygame.image.load(os.path.join(sys.path[0], 'pionowe.png'))
    assert check_crossing_lines(screen, background, begin, end, colorscr, colorbgd) == answer

@pytest.mark.parametrize('players', [0, 1, 2, 6])
def test_init_cars_rand(players):
    """Sprawdza, czy metoda prawidłowo inicjalizuje pojazdy o ilości players"""
    cars = init_cars_rand(players) # lista samochodów
    assert len(cars) == players

src/paperace/car.py
from paperace.const import WIDTH, HEIGHT


class Car:
    """Obiekt car reprezentujący pojazd gracza.
    Funkcje: get_allowed_positions, move
    Atrybuty: xpos, ypos, xv, yv, color"""
    def __init__(self, xpos, ypos, color):
        self.xpos = xpos
        self.ypos = ypos
        self.xv = 0
        self.yv = 0
        self.color = color

    def get_allowed_positions(self):
        """Zwraca listę wszystkich 9-ciu punktów (lub mniej), na jakie może się
        przesunąć pojazd"""
        pos = []
        for y in [-20, 0, 20]:
            for x in [-20, 0, 20]:
                cx = self.xpos+self.xv+x
                cy = self.ypos+self.yv+y
                if cx >= 0 and cx <= WIDTH and cy >= 0 and cy <= HEIGHT:
                    pos.append(tuple((cx, cy)))
        return pos

    def move(self, destination):
        """Przesuwa pojazd gracza na zadaną pozycję (destination). Modyfikuje
        wektor prędkości."""
        self.xv = destination[0] - self.xpos
        self.yv = destination[1] - self.ypos
        self.xpos = destination[0]
        self.ypos = destination[1]


src/paperace/skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.py:

    [console_scripts]
    fibonacci = paperace.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""
from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging

from paperace import __version__

__author__ = "Mariusz Murawski"
__copyright__ = "Mariusz Murawski"
__license__ = "mit"

_logger = logging.getLogger(__name__)


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonnaci demonstration")
    parser.add_argument(
        '--version',
        action='version',
        version='paperace {ver}'.format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


src/paperace/plansza/plansza.py
import pygame
from math import pi

pygame.init()

BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
RED = (255, 0, 0)
GRAY = (47, 79, 79)

WIDTH = 800
HEIGHT = 600
SIZE = 20 # rozmiar kratki

img = pygame.Surface((WIDTH, HEIGHT))
img.fill(WHITE)

# Linie poziome i pionowe
for i in range(SIZE, WIDTH, SIZE):
    pygame.draw.line(img, GRAY, [i, 0], [i, HEIGHT])
for j in range(SIZE, HEIGHT, SIZE):
    pygame.draw.line(img, GRAY, [0, j], [WIDTH, j])

# Linia startu
pygame.draw.line(img, RED, [400, 20], [400, 160], 3)

# Zewnętrzna banda
pygame.draw.arc(img, BLACK, [20, 20, 560, 560], pi/2, 3*pi/2, 3)
pygame.draw.arc(img, BLACK, [WIDTH-20-560, 20, 560, 560], 3*pi/2, pi/2, 3)
pygame.draw.line(img, BLACK, [300, 20], [WIDTH-20-560/2+10, 20], 3)
pygame.draw.line(img, BLACK, [300-11, 580], [WIDTH-20-560/2, 580], 3)

# Wewnętrzna banda
pygame.draw.arc(img, BLACK, [160, 160, 280, 280], pi/2, 3*pi/2, 3)
pygame.draw.arc(img, BLACK, [WIDTH-20-420, 160, 280, 280], 3*pi/2, pi/2, 3)
pygame.draw.line(img, BLACK, [300, 160], [500+5, 160], 3)
pygame.draw.line(img, BLACK, [300-5, 440], [500, 440], 3)

pygame.image.save(img, 'plansza.png')

pygame.quit()

src/paperace/__main__.py
import sys
import os
import pygame
import math
import random
from paperace.car import Car
from paperace.const import *


def main():
    run = True
    show_menu = True
    while run:
        if show_menu:
            human_players = menu()
        after_race_event = race_loop(human_players)
        show_menu = end_game(after_race_event)
    pygame.quit()


def menu():
    """Wyświetla menu, w którym można wybrać ilość graczy lub zamknąć grę.
    Zwraca ilość graczy."""
    screen.blit(background, background.get_rect())
    game_font = pygame.font.SysFont("monospace", 30)
    game_font.set_bold(True)
    text_to_show = game_font.render("1 Player", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2-120)))
    text_to_show = game_font.render("2 Players", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2-80)))
    text_to_show = game_font.render("3 Players", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2-40)))
    text_to_show = game_font.render("4 Players", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2)))
    text_to_show = game_font.render("5 Players", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2+40)))
    text_to_show = game_font.render("6 Players", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2+80)))
    text_to_show = game_font.render("Exit", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2+120)))
    pygame.display.flip()
    
    menu_loop = True
    while menu_loop:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()
            if event.type == pygame.MOUSEBUTTONDOWN:
                mouseclick = pygame.mouse.get_pos()
                if mouseclick[0] > 320 and mouseclick[0] < 480 and mouseclick[1] > 160 and mouseclick[1] < 200:
                    return 1
                elif mouseclick[0] > 320 and mouseclick[0] < 480 and mouseclick[1] > 200 and mouseclick[1] < 240:
                    return 2
                elif mouseclick[0] > 320 and mouseclick[0] < 480 and mouseclick[1] > 240 and mouseclick[1] < 280:
                    return 3
                elif mouseclick[0] > 320 and mouseclick[0] < 480 and mouseclick[1] > 280 and mouseclick[1] < 320:
                    return 4
                elif mouseclick[0] > 320 and mouseclick[0] < 480 and mouseclick[1] > 320 and mouseclick[1] < 360:
                    return 5
                elif mouseclick[0] > 320 and mouseclick[0] < 480 and mouseclick[1] > 360 and mouseclick[1] < 400:
                    return 6
                elif mouseclick[0] > 360 and mouseclick[0] < 440 and mouseclick[1] > 400 and mouseclick[1] < 440:
                    pygame.quit()
                    sys.exit()


def race_loop(human_players):
    """Przyjmuje ilość graczy. Ustawia wszystkie parametry oraz wykonuje główną
    pętlę gry. Zwraca wydarzenie: zwycięstwo, wyjście poza trasę, kraksa."""
    cars = init_cars_rand(human_players)  # Lista pojazdów
    cp = 0  # Aktualny gracz na liście - Current Player

    screen.blit(background, background.get_rect())
    for car in cars:
        pygame.draw.circle(screen, car.color, [car.xpos, car.ypos], 5)
    pygame.draw.rect(screen, cars[0].color, (0, 0, WIDTH, HEIGHT), 3)
    pygame.display.flip()

    run = True
    while run:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                sys.exit()
            if event.type == pygame.MOUSEBUTTONDOWN:
                nearestpoint = get_nearest_point(pygame.mouse.get_pos())
                if nearestpoint in get_allowed_car_positions(cars, cp):
                    xoldpos = cars[cp].xpos
                    yoldpos = cars[cp].ypos
                    cars[cp].move(nearestpoint)
                    pygame.draw.line(screen, cars[cp].color, [xoldpos, yoldpos], [cars[cp].xpos, cars[cp].ypos], 2)
                    if check_crossing_lines(background, screen, (cars[cp].xpos, cars[cp].ypos), (xoldpos, yoldpos), BLACK, cars[cp].color):
                        pygame.draw.circle(screen, cars[cp].color, [cars[cp].xpos, cars[cp].ypos], 5)
                        del cars[cp]
                        cp -= 1
                        if len(cars) == 0:
                            return "Crash"
                    if check_crossing_lines(background, screen, (cars[cp].xpos, cars[cp].ypos), (xoldpos, yoldpos), RED, cars[cp].color) and xoldpos != 400:
                        pygame.draw.circle(screen, cars[cp].color, [cars[cp].xpos, cars[cp].ypos], 5)
                        return "Winning"
                    if len(cars[cp].get_allowed_positions()) == 0:
                        pygame.draw.circle(screen, cars[cp].color, [cars[cp].xpos, cars[cp].ypos], 5)
                        del cars[cp]
                        cp -= 1
                        if len(cars) == 0:
                            return "Out"
                    pygame.draw.circle(screen, cars[cp].color, [cars[cp].xpos, cars[cp].ypos], 5)
                    
                    cp = cp + 1
                    if cp >= len(cars):
                        cp = 0

                    pygame.draw.rect(screen, cars[cp].color, (0, 0, WIDTH, HEIGHT), 3)
        pygame.display.flip()


def end_game(after_race_event):
    """Przyjmuje wydarzenie: koniec gry, wyjście poza trasę, kraksa. Wyświetla
    wynik wyścigu oraz pyta o dalsze kroki: menu, nowa gra (ta sama liczba
    graczy i nowe ustawienia), wyjście. Zwraca True jeśli trzeba wyświetlić
    menu."""
    game_font = pygame.font.SysFont("monospace", 30)
    game_font.set_bold(True)
    if after_race_event == "Crash":
        text_to_show = game_font.render("Crash!", False, RED, WHITE)
        screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2-40)))
    elif after_race_event == "Winning":
        text_to_show = game_font.render("Winning!", False, RED, WHITE)
        screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2-40)))
    elif after_race_event == "Out":
        text_to_show = game_font.render("Next move out of screen!", False, RED, WHITE)
        screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2-40)))
    text_to_show = game_font.render("Play again", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2+40)))
    text_to_show = game_font.render("Menu", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2+80)))
    text_to_show = game_font.render("Exit", False, WHITE, BLACK)
    screen.blit(text_to_show, text_to_show.get_rect(center=(WIDTH/2, HEIGHT/2+120)))
    pygame.display.flip()
    
    menu_loop = True
    while menu_loop:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()
            if event.type == pygame.MOUSEBUTTONDOWN:
                mouseclick = pygame.mouse.get_pos()
                if mouseclick[0] > 300 and mouseclick[0] < 500 and mouseclick[1] > 320 and mouseclick[1] < 360:
                    return False
                elif mouseclick[0] > 360 and mouseclick[0] < 440 and mouseclick[1] > 360 and mouseclick[1] < 400:
                    return True
                elif mouseclick[0] > 360 and mouseclick[0] < 440 and mouseclick[1] > 400 and mouseclick[1] < 440:
                    pygame.quit()
                    sys.exit()


def init_cars_rand(players):
    """Zwraca listę pojazdów (cars) w zależności od liczby graczy (players).
    Pojazdy ustawiane są losowo na linii ([400, 40], [400, 140])."""
    cars = []
    colors = [RED, GREEN, BLUE, YELLOW, CYAN, MAGENTA]
    y_positions = [40, 60, 80, 100, 120, 140]
    for _ in range(players):
        y = y_positions[random.randint(0, len(y_positions)-1)]
        y_positions.remove(y)
        color = colors[random.randint(0, len(colors)-1)]
        colors.remove(color)
        cars.append(Car(400, y, color))
    return cars


def get_nearest_point(mousexy):
    """Zwraca punkt najbliższego przecięcia kratek po kliknięciu myszką."""
    if mousexy[0] % 20 < 10:
        x = int(mousexy[0] / 20.0) * 20
    else:
        x = int(math.ceil(mousexy[0] / 20.0)) * 20
    if mousexy[1] % 20 < 10:
        y = int(mousexy[1] / 20.0) * 20
    else:
        y = int(math.ceil(mousexy[1] / 20.0)) * 20
    return (x, y)


def get_allowed_car_positions(cars, i):
    """Zwraca listę punktów, na które może przemieścić się samochód cars[i]
    oraz usuwa pozycje, na których znajduje się pojazd przeciwnika)."""
    positions = cars[i].get_allowed_positions()
    for c in cars:
        position_to_remove = ((c.xpos, c.ypos))
        if position_to_remove in positions:
            positions.remove(position_to_remove)
    return positions


def check_crossing_lines(first_img, second_img, rectbegin, rectend, first_color, second_color):
    """Zwraca True jeśli na prostokątnym obszarze (rectbegin) X (rectend) na
    obrazkach first_img i second_img przecinają się linie o kolorach
    odpowiednio: first_color i second_color."""
    # Ustawia prawidłowy prostokąt LG X PD
    xbegin = min(rectbegin[0], rectend[0])
    ybegin = min(rectbegin[1], rectend[1])
    xend = max(rectbegin[0], rectend[0])
    yend = max(rectbegin[1], rectend[1])
    if xbegin == xend:
        xend = xend + 1
    if ybegin == yend:
        yend = yend + 1

    for y in range(ybegin, yend):
        for x in range(xbegin, xend):
            first_img_color = first_img.get_at((x, y))
            second_img_color = second_img.get_at((x, y))
            if first_img_color == first_color and second_img_color == second_color:
                return True
    return False


if __name__ == '__main__':
    pygame.init()
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("PapeRace")

    imgfile = os.path.join(os.path.split(__file__)[0], 'plansza\plansza.png')
    background = pygame.image.load(imgfile).convert()
    
    main()
